File: src/main/java/io/github/slimefunguguproject/bump/core/attributes/CooldownItem.java
package io.github.slimefunguguproject.bump.core.attributes;

import javax.annotation.Nonnull;

import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import io.github.slimefunguguproject.bump.utils.ValidateUtils;
import io.github.slimefunguguproject.bump.utils.constant.Keys;
import io.github.thebusybiscuit.slimefun4.core.attributes.ItemAttribute;
import io.github.thebusybiscuit.slimefun4.libraries.dough.data.persistent.PersistentDataAPI;

/**
 * This {@link ItemAttribute} indicates that this item has cooldown time.
 *
 * @author ybw0014
 */
public interface CooldownItem extends ItemAttribute {
    /**
     * This method returns the cooldown time in seconds.
     *
     * @return cooldown time in seconds.
     */
    int getCooldown();

    default void setCooldown(@Nonnull ItemStack itemStack) {
        if (ValidateUtils.noAirItem(itemStack)) {
            ItemMeta im = itemStack.getItemMeta();
            PersistentDataAPI.setLong(im, Keys.LAST_USED, System.currentTimeMillis());
            itemStack.setItemMeta(im);
        }
    }

    /**
     * This method checks if given {@link ItemStack} has cooled down.
     *
     * @param itemStack The {@link ItemStack} to be checked
     *
     * @return if the item can be used now
     */
    default boolean isCooldown(@Nonnull ItemStack itemStack) {
        if (ValidateUtils.noAirItem(itemStack)) {
            ItemMeta im = itemStack.getItemMeta();

            if (PersistentDataAPI.hasLong(im, Keys.LAST_USED)) {
                long lastUsed = PersistentDataAPI.getLong(im, Keys.LAST_USED);

                return lastUsed + getCooldown() * 1000L < System.currentTimeMillis();
            } else {
                return true;
            }
        } else {
            return false;
        }
    }

    default boolean checkCooldown(@Nonnull ItemStack itemStack) {
        if (ValidateUtils.noAirItem(itemStack)) {
            if (isCooldown(itemStack)) {
                setCooldown(itemStack);
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/core/attributes/CooldownItem.kt
package io.github.slimefunguguproject.bump.core.attributes

import io.github.slimefunguguproject.bump.utils.constant.Keys
import io.github.slimefunguguproject.bump.utils.items.ValidateUtils.noAirAndHasItemMeta
import io.github.thebusybiscuit.slimefun4.core.attributes.ItemAttribute
import io.github.thebusybiscuit.slimefun4.libraries.dough.data.persistent.PersistentDataAPI
import org.bukkit.inventory.ItemStack

interface CooldownItem : ItemAttribute {
    companion object {
        private const val INVALID_ITEMSTACK = "Invalid ItemStack."
    }

    /**
     * This variable returns the cooldown time in seconds.
     */
    fun getCooldown(): Int

    /**
     * Set the [ItemStack] to cooldown.
     *
     * @param itemStack The [ItemStack] to be set to cooldown.
     */
    fun setCooldown(itemStack: ItemStack) {
        require(noAirAndHasItemMeta(itemStack)) { INVALID_ITEMSTACK }

        val im = itemStack.itemMeta!!
        PersistentDataAPI.setLong(im, Keys.LAST_USED, System.currentTimeMillis())
        itemStack.setItemMeta(im)
    }

    /**
     * This method checks if given [ItemStack] has cooled down.
     *
     * @param itemStack The [ItemStack] to be checked
     *
     * @return if the item can be used now
     */
    fun isCooldown(itemStack: ItemStack): Boolean {
        require(noAirAndHasItemMeta(itemStack)) { INVALID_ITEMSTACK }

        val im = itemStack.itemMeta!!

        return if (PersistentDataAPI.hasLong(im, Keys.LAST_USED)) {
            val lastUsed = PersistentDataAPI.getLong(im, Keys.LAST_USED)

            lastUsed + getCooldown() * 1000L < System.currentTimeMillis()
        } else {
            true
        }
    }

    /**
     * This method checks if given [ItemStack] has cooled down,
     * if so, set the [ItemStack] to be just used.
     *
     * @param itemStack The [ItemStack] to be checked
     *
     * @return if the item can be used now
     */
    fun checkCooldown(itemStack: ItemStack): Boolean {
        require(noAirAndHasItemMeta(itemStack)) { INVALID_ITEMSTACK }

        return if (isCooldown(itemStack)) {
            setCooldown(itemStack)
            true
        } else {
            false
        }
    }
}
